### ISSUE TITLE
Release build crashes on Android & iOS (On Launch)

### ISSUE THREAD

**Issue by YogeshYo** [facebook/react-native#18121]
### Is this a bug report?

Yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment
    OS:  macOS Sierra 10.13.2
    Node:  8.4.0
    Yarn:  1.0.2
    npm:  5.6.0
    Watchman:  4.9.0
    Xcode:  Xcode  9 Build version 8E3004b
    Android Studio: 3 AI-162.4069837

  Packages: (wanted => installed)
    react-native: 0.52.0 => 0.52.0
    react: 16.2.0 => 16.2.0

  Target Platform: iOS (10.3, Android 7.1.2

### Steps to Reproduce

(Write your steps here:)

iOS & Android release build crashes on application launch, debug build works fine.
![2336165327667826946](https://user-images.githubusercontent.com/33345378/36749393-a061790e-1c20-11e8-8671-039ea956457e.png)

**Log**
02-27 21:41:13.822  8695  8754 E AndroidRuntime: FATAL EXCEPTION: mqt_native_modules
02-27 21:41:13.822  8695  8754 E AndroidRuntime: Process: com.reactwork, PID: 8695
02-27 21:41:13.822  8695  8754 E AndroidRuntime: com.facebook.react.common.JavascriptException: undefined is not a function (evaluating '(void 0)()'), stack:
02-27 21:41:13.822  8695  8754 E AndroidRuntime: <unknown>@531:415
02-27 21:41:13.822  8695  8754 E AndroidRuntime: c@2:837
02-27 21:41:13.822  8695  8754 E AndroidRuntime: n@2:409
02-27 21:41:13.822  8695  8754 E AndroidRuntime: t@2:262
02-27 21:41:13.822  8695  8754 E AndroidRuntime: <unknown>@402:1044
02-27 21:41:13.822  8695  8754 E AndroidRuntime: c@2:837
02-27 21:41:13.822  8695  8754 E AndroidRuntime: n@2:409
02-27 21:41:13.822  8695  8754 E AndroidRuntime: t@2:262
02-27 21:41:13.822  8695  8754 E AndroidRuntime: <unknown>@336:133
02-27 21:41:13.822  8695  8754 E AndroidRuntime: c@2:837
02-27 21:41:13.822  8695  8754 E AndroidRuntime: n@2:409
02-27 21:41:13.822  8695  8754 E AndroidRuntime: t@2:262
02-27 21:41:13.822  8695  8754 E AndroidRuntime: <unknown>@303:161
02-27 21:41:13.822  8695  8754 E AndroidRuntime: c@2:837
02-27 21:41:13.822  8695  8754 E AndroidRuntime: n@2:409
02-27 21:41:13.822  8695  8754 E AndroidRuntime: t@2:262

### Expected Behavior

The release build should not crash.

### Actual Behavior

The debug build working fine.
The release build crashes on launch after displaying native splash screen.

### Reproducible Demo
https://github.com/Tradesmen/CRASH/tree/master.

(Paste the link to an example project and exact instructions to reproduce the issue.)

**Comment by joshyhargreaves** [comment]
@YogeshYo are you able to provide us a reproducible demo?

**Comment by YogeshYo** [comment]
What demo do you need on this? Please guide. I have attached screenshot. The crash takes place as soon as we launch application even before appearing splash screen. Do you need a package.json details?

**Comment by YogeshYo** [comment]
I tried upgrading react-native & reinstalling node modules.

**Comment by joshyhargreaves** [comment]
Ideally we’d have a project you’ve uploaded which exhibits this crash; maybe you can reproduce with your depdencies and a new project.

However, when I’ve seen crashes that happen on both iOS and Android at startup, it usually points to a problem that only occurs when building the jsbundle in release mode.

To test this theory, modify your JSCodeCodeLocation on Android or iOS to point to: http://localhost:8081/index.bundle?platform=ios&dev=false&minify=false and see if the stack trace is any more informative.

**Comment by YogeshYo** [comment]
@joshyhargreaves, i tried above   for android protected String getJSMainModuleName() {
      return "http://localhost:8081/index.bundle?platform=android&dev=false&minify=false";
    }

**Comment by joshyhargreaves** [comment]
The actual point is finding the cause of the crash!

**Comment by YogeshYo** [comment]
protected String getJSMainModuleName() {
return "http://localhost:8081/index.bundle?platform=android&dev=false&minify=false";
}
is this right in mainapplication.java?

**Comment by YogeshYo** [comment]
@joshyhargreaves, with above I got the attached message.
 I already tried removing & relinking Datepicker library. I will try again relinking datepicker with a different approach.
Please let me know your inputs on same. am I moving in right direction?

**Comment by YogeshYo** [comment]
@joshyhargreaves, I removed Datepicker library but no difference. 
 {
  "name": "ReactWork",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "node node_modules/react-native/local-cli/cli.js start",
    "test": "jest"
  },
  "dependencies": {
    "moment": "^2.20.1",
    "react": "16.2.0",
    "react-native": "0.53.3",
    "react-native-android-keyboard-adjust": "^1.1.1",
    "react-native-extra-dimensions-android": "^0.21.0",
    "react-native-maps": "^0.19.0",
    "react-native-pages": "^0.6.1",
    "react-native-root-siblings": "^3.0.0",
    "react-navigation": "^1.0.0-beta.27",
    "react-redux": "^5.0.6",
    "redux": "^3.7.2",
    "redux-persist": "^4.10.1",
    "redux-saga": "^0.15.6"
  },
  "devDependencies": {
    "babel-jest": "22.0.4",
    "babel-preset-react-native": "4.0.0",
    "jest": "22.0.5",
    "react-test-renderer": "16.2.0"
  },
  "jest": {
    "preset": "react-native"
  },
  "rnpm": {
    "assets": [
      "./assets/fonts"
    ]
  }
}

I was able to generate release build earlier but now I am stuck to this from last 4-5 days. am I following your solution to get the message for crash correctly?

**Comment by YogeshYo** [comment]
@joshyhargreaves, With  jsCodeLocation = [NSURL URLWithString:@"http://localhost:8081/index.bundle?platform=ios&dev=false&minify=false"];. I am able to get tha same message as i posted earlier.
![2336165327667826946](https://user-images.githubusercontent.com/33345378/36993272-65664dea-20d3-11e8-8776-5d6c7667f99a.png)

**Comment by YogeshYo** [comment]
@joshyhargreaves, anything else I can try? Release build was working fine until last week. I tried a release build with removed lib added since then but same error on release build. "App crashes after splash screen on both Android & iOS"

**Comment by joshyhargreaves** [comment]
@yogesho, the thing I might suggest, is if you can keep your native dependencies, remove any application code that you don’t want to upload, and create a repro project, then that’s be easier for one of us to take a look.

**Comment by YogeshYo** [comment]
@joshyhargreaves, can you please share you email id so that i can share codes to you

**Comment by YogeshYo** [comment]
@joshyhargreaves, can you please look into this? I tried to reach you on Gmail. Thanks

**Comment by YogeshYo** [comment]
Hi Josh,

I will do this still I think if you have a chance to have at least one look then I am sure you can at least give me hint to move ahead on this. Believe me, I tried everything I can & now I am stuck.
It's already around 10 days so I am requesting you to have one look.

Thanks

**Comment by YogeshYo** [comment]
HI @joshyhargreaves, reproducible demo added
https://github.com/Tradesmen/CRASH/tree/master.
I request you to remove "Need more information" badge.
Thanks

**Comment by testshallpass** [comment]
@YogeshYo I have similar issue and it start happening after upgrade to react-native 0.50.4. I'm told the cause could be a that a dependency violates things that because invalid in react 16.0.0.

**Comment by Tradesmen** [comment]
@testshallpass, I went line by line & found two culprits for this issue (i.e only in release build).
1. Platform.select({
    "ios": () => StatusBar.setBarStyle("light-content"),
    "android": () => StatusBar.setBackgroundColor("#263238")
})();

The second was https://github.com/facebook/react-native/issues/4831 for which I proper message in the crash log which helped me to fix.

Still, i think for the 1st either this should be rectified in debug build or at least get a proper error message in the crash log.

**Comment by YogeshYo** [comment]
@testshallpass, I went line by line & found two culprits for this issue (i.e only in release build).

Platform.select({
"ios": () => StatusBar.setBarStyle("light-content"),
"android": () => StatusBar.setBackgroundColor("#263238")
})();
The second was #4831 for which I got a proper message in the crash log which helped me to fix.

Still, i think for the 1st; either the issue should be rectified in debug build or at least get a proper error message in the crash log.
